client/my-sites/plan-features-2023-grid/index.tsx

```diff
@@ -852,26 +865,76 @@ export class PlanFeatures2023Grid extends Component<
 			} );
 	}
 
+	renderStorageAddOnDropdown( properties: PlanProperties ) {
+		const { planName, storageOptions, storageFeatures } = properties;
+		const { selectedStorage } = this.state;
+
+		return (
+			<SelectDropdown
+				onToggle={ ( { open: isOpen }: { open: boolean } ) => {
+					if ( isOpen ) {
+						// record interaction event here
+					}
+				} }
+				selectedText={
+					selectedStorage[ planName ]
+						? getStorageStringFromFeature( selectedStorage[ planName ] )
+						: getStorageStringFromFeature( storageFeatures[ 0 ] )
+				}
+			>
+				{ storageOptions.map( ( storageOption ) => {
+					return (
+						<SelectDropdown.Item
+							key={ `${ planName } ${ storageOption }` }
+							selected={ storageOption === selectedStorage[ planName ] }
+							onClick={ () => {
+								const updatedSelectedStorage = {
+									...selectedStorage,
+									[ planName ]: storageOption,
+								};
+
+								this.setState( {
+									...this.state,
+									selectedStorage: updatedSelectedStorage,
+								} );
+							} }
+						>
+							{ getStorageStringFromFeature( storageOption ) }
+						</SelectDropdown.Item>
+					);
+				} ) }
+			</SelectDropdown>
+		);
+	}
+
 	renderPlanStorageOptions( planPropertiesObj: PlanProperties[], options?: PlanRowOptions ) {
-		const { translate } = this.props;
+		const { translate, intervalType } = this.props;
+
 		return planPropertiesObj
 			.filter( ( { isVisible } ) => isVisible )
 			.map( ( properties ) => {
 				if ( options?.isMobile && isWpcomEnterpriseGridPlan( properties.planName ) ) {
 					return null;
 				}
 
-				const { planName, storageOptions } = properties;
-				const storageJSX = storageOptions.map( ( storageFeature: string ) => {
-					if ( storageFeature.length <= 0 ) {
-						return;
-					}
-					return (
-						<div className="plan-features-2023-grid__storage-buttons" key={ planName }>
-							{ getStorageStringFromFeature( storageFeature ) }
-						</div>
-					);
-				} );
+				const { planName, storageOptions, storageFeatures } = properties;
+				const canUpgradeStorageForPlan =
+					storageOptions.length > 1 &&
+					intervalType === 'yearly' &&
+					isEnabled( 'plans/upgradeable-storage' );
+
+				const storageJSX = canUpgradeStorageForPlan
+					? this.renderStorageAddOnDropdown( properties )
+					: storageFeatures.map( ( storageFeature: string ) => {
+							if ( storageFeature.length <= 0 ) {
+								return;
+							}
```

Comment:
I don't understand why this is necessary, but probably need a closer read. `storageFeatures` should either have expected values or not 🤔 

also, can string length be negative? :D

Reply:
> also, can string length be negative? :D

😆  

> I don't understand why this is necessary, but probably need a closer read. storageFeatures should either have expected values or not 🤔

I didn't quite understand either -- it's something I carried over from the existing code https://github.com/Automattic/wp-calypso/pull/79041/files/da55eb56ec9bc6270604f127ce1f3542936b74f1#diff-e268185bf1c246017f6796f5f74ff3c0ae49b12e2acf525d8074859449573a19L866 

I didn't look too deeply into the git history because I was spending time wrapping my head around the conventions in this file, but now that you mention it, it's probably worth double checking

Reply:
So I played around a bit more. Initialization with the `forEach` was _not necessary_, and I conflated it with another change that fixed the typescript error I was seeing.

What matters is actually initializing state in the constructor, rather than as a class member

https://github.com/Automattic/wp-calypso/pull/79041/files#diff-e268185bf1c246017f6796f5f74ff3c0ae49b12e2acf525d8074859449573a19L245-R255

I'm not quite sure what to make of it, or if moving state initialization to the constructor affects any behavior. 🤔

Reply:
Traced the change back to [this PR](https://github.com/Automattic/wp-calypso/pull/71378), but I don't see any clear reasoning as to why this exists.

I'm assuming it was an oversight, so I'll change or remove the condition

Reply:
[0300cd1](https://github.com/Automattic/wp-calypso/pull/79041/commits/0300cd19e8fa634572522974399363ebcde56637)